README.md
# Haproxy in Docker
Alfresco-specific haproxy docker images.

## Overview

This is the repository building Haproxy Docker images. At the moment there are 3 templates available:

* a front proxy
* a load balancer between alfresco and solr
* a load balancer between solr and alfresco

An init script choses the right template based on an environment variable and creates the final haproxy configuration file. Templates are themselves scripts, to allow for further tweaking, for example adding a new kind of backend. 

All images are automatically built by [jenkins-2](https://jenkins-2.xenit.eu) and published to [hub.xenit.eu/public/haproxy-alfresco](https://hub.xenit.eu/public/haproxy-alfresco).

## Environment variables

There are several environment variables available to tweak the behaviour.

| Variable                         |  Default                                          | Comments                                                                       |
| -------------------------------- | ------------------------------------------------- | ------------------------------------------------------------------------------ |
| HAPROXY_TEMPLATE                 | template-front-proxy-alfresco.sh                  | Type of proxy                                                                  |
| HAPROXY_HOSTNAME                 | localhost                                         | Hostname to be sent to a syslog/logstash server.                                          |
| HAPROXY_SEND_LOGS                | false                                             | Whether to send logs to a syslog/logstash server.                                       |
| HAPROXY_LOGS_SERVER              | 10.88.10.50:514                                   | Hetzner syslog/logstash server.                                |
| HAPROXY_STATS_TIMEOUT            | 2m                                                | Stats timeout.                                                                 |
| HAPROXY_COOKIE                   | JSESSIONID                                        | Cookie to be used. Valid for front proxy.                                      |
| HAPROXY_BALANCE                  | hdr(X-Forwarded-For)/leastconn                    | Balancing method. Different defaults for front proxy/alfresco-solr-alfresco-lb |
| HAPROXY_TIMEOUT_CLIENT           | 30m                                               |                                                                                |
| HAPROXY_TIMEOUT_CONNECT          | 4s                                                |                                                                                |
| HAPROXY_TIMEOUT_SERVER           | 30m                                               |                                                                                |
| HAPROXY_TIMEOUT_CHECK            | 5s                                                |                                                                                |
| HAPROXY_INCLUDE_FRONTEND_STATS   | true                                              | Whether to generate statistics to be used by Prometheus scraping.              |
| HAPROXY_FRONTEND_STATS           | stats                                             | Frontend name for statistics.                                                  |
| HAPROXY_FRONTEND_INTERNET        | internet                                          | Frontend name for internet. Valid for front proxy.                             |
| HAPROXY_FRONTEND_INTERNET_BIND   | *:80                                              | Bind for frontend internet. Valid for front proxy.                             |
| HAPROXY_FRONTEND_ALFRESCO        | alfresco                                          | Frontend name for alfresco. Valid for lb-alfresco-solr.                        |
| HAPROXY_FRONTEND_ALFRESCO_BIND   | *:8882                                            | Bind for frontend alfresco. Valid for lb-alfresco-solr.                        |
| HAPROXY_FRONTEND_SOLR            | solr                                              | Frontend name for solr. Valid for lb-solr-alfresco.                        |
| HAPROXY_FRONTEND_SOLR_BIND       | *:8880                                            | Bind for frontend solr. Valid for lb-solr-alfresco.                        |
| HAPROXY_BACKEND_ALFRESCO         | alfresco                                          | Backend name for alfresco. Valid for front proxy and lb-solr-alfresco.         |
| HAPROXY_BACKEND_ALFRESCO_CHECK   | /alfresco/s/api/server                            | Check for alfresco backend. Alternative: /actuators/health. Valid for front proxy and lb-solr-alfresco.                    |
| HAPROXY_BACKEND_ALFRESCO_COUNT   | 1                                                 | Number of alfresco backends. Valid for front proxy and lb-solr-alfresco.                                                   |
| HAPROXY_SERVICE_ALFRESCO         | alfresco                                          | Alfresco service name (docker DNS). Valid for front proxy and lb-solr-alfresco.                                            |
| HAPROXY_INCLUDE_SHARE            | false                                             | Whether to include share as a backend. Valid for front proxy.                  |
| HAPROXY_BACKEND_SHARE            | share                                             | Backend name for share. Valid for front proxy.                                 |
| HAPROXY_BACKEND_SHARE_CHECK      | share                                             | Check for share backend. Valid for front proxy.                                                       |
| HAPROXY_BACKEND_SHARE_COUNT      | 1                                                 | Number of share backends. Valid for front proxy.                                                     |
| HAPROXY_SERVICE_SHARE            | share                                             | Share service name (docker DNS). Valid for front proxy.                                               |
| HAPROXY_INCLUDE_WORKSPACE        | false                                             | Whether to include digital workspace as a backend. Valid for front proxy.      |
| HAPROXY_BACKEND_WORKSPACE        | workspace                                         | Backend name for digital workspace. Valid for front proxy.                     |
| HAPROXY_BACKEND_WORKSPACE_CHECK  | workspace                                         | Check for workspace backend. Valid for front proxy.                                                    |
| HAPROXY_BACKEND_WORKSPACE_COUNT  | 1                                                 | Number of workspace backends. Valid for front proxy.                                                 |
| HAPROXY_SERVICE_WORKSPACE        | digital-workspace                                 | Workspace service name (docker DNS). Valid for front proxy.                                            |
| HAPROXY_INCLUDE_FINDER           | false                                             | Whether to include finder as a backend. Valid for front proxy.                 |
| HAPROXY_BACKEND_FINDER           | finder                                            | Backend name for finder. Valid for front proxy.                                |
| HAPROXY_BACKEND_FINDER_CHECK     | /                                                 | Check for finder backend. Valid for front proxy.                                                     |
| HAPROXY_BACKEND_FINDER_COUNT     | 1                                                 | Number of finder backends. Valid for front proxy.                                                    |
| HAPROXY_SERVICE_FINDER           | alfred-finder                                     | Finder service name (docker DNS). Valid for front proxy.                                             |
| HAPROXY_INCLUDE_OOI              | false                                             | Whether to include ooi service as a backend. Valid for front proxy.                 |
| HAPROXY_BACKEND_OOI              | ooi-service                                       | Backend name for ooi service. Valid for front proxy.                                |
| HAPROXY_BACKEND_OOI_CHECK        | /                                                 | Check for ooi service backend. Valid for front proxy.                                                     |
| HAPROXY_BACKEND_OOI_COUNT        | 1                                                 | Number of ooi service backends. Valid for front proxy.                                                    |
| HAPROXY_SERVICE_OOI              | alfresco-ooi-service                              | Ooi service name (docker DNS). Valid for front proxy.                                             |
| HAPROXY_BACKEND_SOLR             | solr                                              | Backend name for solr. Valid for lb-alfresco-solr.                             |
| HAPROXY_BACKEND_SOLR_CHECK       | /solr4                                            | Check for solr backend. Valid for lb-alfresco-solr.                             |
| HAPROXY_BACKEND_SOLR_COUNT       | 1                                                 | Number of solr backends.  Valid for lb-alfresco-solr.                          |
| HAPROXY_SERVICE_SOLR             | solr                                              | Solr service name (docker DNS).                                               |
| HAPROXY_RESOLVE_RETRIES          | 3                                                 |                                                                                |
| HAPROXY_TIMEOUT_RESOLVE          | 1s                                                |                                                                                |
| HAPROXY_TIMEOUT_RETRY            | 1s                                                |                                                                                |
| HAPROXY_HOLD_OTHER               | 10s                                               |                                                                                |
| HAPROXY_HOLD_REFUSED             | 10s                                               |                                                                                |
| HAPROXY_HOLD_NX                  | 10s                                               |                                                                                |
| HAPROXY_HOLD_TIMEOUT             | 10s                                               |                                                                                |
| HAPROXY_HOLD_VALID               | 10s                                               |                                                                                |
| HAPROXY_HOLD_OBSOLETE            | 10s                                               |                                                                                |

**Maintained by:**

Roxana Angheluta <roxana.angheluta@xenit.eu>

### How to build

To build a local version of the _haproxy_ image:

```bash
./gradlew buildDockerImage
```

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), with additional info about the chronology things are added/fixed/changed and - where possible - links to the PRs involved.

## [Unreleased - 0.0.1]

### Added
* [2021-04-22] [PR #2](https://github.com/xenit-eu/docker-haproxy-alfresco/pull/2) Added support for ooi service backend.
* [2021-02-05] [PR #1](https://github.com/xenit-eu/docker-haproxy-alfresco/pull/1) Initial version.	


src/main/resources/templates/template-lb-alfresco-solr.sh
#!/bin/sh

set -e

HAPROXY_SEND_LOGS=${HAPROXY_SEND_LOGS:-false}
HAPROXY_INCLUDE_FRONTEND_STATS=${HAPROXY_INCLUDE_FRONTEND_STATS:-true}

########## global
echo "global
  log-send-hostname ${HAPROXY_HOSTNAME:-localhost}
  log stdout format raw local0 debug
  maxconn 4096
  nbthread 1
  uid 99
  gid 99
  stats socket /var/run/haproxy.sock mode 600 level admin
  stats timeout ${HAPROXY_STATS_TIMEOUT:-2m}"

if [[ $HAPROXY_SEND_LOGS = 'true' ]]
then
    echo "
  log ${HAPROXY_LOGS_SERVER:-10.88.10.50:514} len 2048 local0"
fi

########## defaults
echo "defaults
  log global
  mode http
  balance ${HAPROXY_BALANCE:-leastconn}
  retries 2
  timeout client ${HAPROXY_TIMEOUT_CLIENT:-30m}
  timeout connect ${HAPROXY_TIMEOUT_CONNECT:-4s}
  timeout server ${HAPROXY_TIMEOUT_SERVER:-30m}
  timeout check ${HAPROXY_TIMEOUT_CHECK:-5s}
  option redispatch
  option httplog"

########## stats frontend
if [[ $HAPROXY_INCLUDE_FRONTEND_STATS = 'true' ]]
then
echo "
frontend ${HAPROXY_FRONTEND_STATS:-stats}
   bind *:8404
   option http-use-htx
   http-request use-service prometheus-exporter if { path /metrics }
   stats enable
   stats uri /stats
   stats refresh 10s"
fi

########## alfresco frontend
echo "
frontend ${HAPROXY_FRONTEND_ALFRESCO:-alfresco}
  bind ${HAPROXY_FRONTEND_ALFRESCO_BIND:-*:8882}
  default_backend ${HAPROXY_BACKEND_SOLR:-solr}"

########## solr backend
echo "
backend ${HAPROXY_BACKEND_SOLR:-solr}
  option httpchk OPTIONS ${HAPROXY_BACKEND_SOLR_CHECK:-/solr4}
  server-template ${HAPROXY_BACKEND_SOLR:-solr}- ${HAPROXY_BACKEND_SOLR_COUNT:-1} ${HAPROXY_SERVICE_SOLR:-solr}:8080 check resolvers docker init-addr libc,none"

########## resolvers
echo "
resolvers docker
  nameserver docker 127.0.0.11:53
  resolve_retries ${HAPROXY_RESOLVE_RETRIES:-3}
  timeout resolve ${HAPROXY_TIMEOUT_RESOLVE:-1s}
  timeout retry   ${HAPROXY_TIMEOUT_RETRY:-1s}
  hold other      ${HAPROXY_HOLD_OTHER:-10s}
  hold refused    ${HAPROXY_HOLD_REFUSED:-10s}
  hold nx         ${HAPROXY_HOLD_NX:-10s}
  hold timeout    ${HAPROXY_HOLD_TIMEOUT:-10s}
  hold valid      ${HAPROXY_HOLD_VALID:-10s}
  hold obsolete   ${HAPROXY_HOLD_OBSOLETE:-10s}" 



src/main/resources/templates/template-lb-solr-alfresco.sh
#!/bin/sh

set -e

HAPROXY_SEND_LOGS=${HAPROXY_SEND_LOGS:-false}
HAPROXY_INCLUDE_FRONTEND_STATS=${HAPROXY_INCLUDE_FRONTEND_STATS:-true}

########## global
echo "
global
  log-send-hostname ${HAPROXY_HOSTNAME:-localhost}
  log stdout format raw local0 debug
  maxconn 4096
  nbthread 1
  uid 99
  gid 99
  stats socket /var/run/haproxy.sock mode 600 level admin
  stats timeout ${HAPROXY_STATS_TIMEOUT:-2m}"

if [[ $HAPROXY_SEND_LOGS = 'true' ]]
then
  echo "
  log ${HAPROXY_LOGS_SERVER:-10.88.10.50:514} len 2048 local0" 
fi

########## defaults
echo "
defaults
  log global
  mode http
  balance ${HAPROXY_BALANCE:-leastconn}
  retries 2
  timeout client ${HAPROXY_TIMEOUT_CLIENT:-30m}
  timeout connect ${HAPROXY_TIMEOUT_CONNECT:-4s}
  timeout server ${HAPROXY_TIMEOUT_SERVER:-30m}
  timeout check ${HAPROXY_TIMEOUT_CHECK:-5s}
  option redispatch
  option httplog"

########## stats frontend
if [[ $HAPROXY_INCLUDE_FRONTEND_STATS = 'true' ]]
then
echo "
frontend ${HAPROXY_FRONTEND_STATS:-stats}
   bind *:8404
   option http-use-htx
   http-request use-service prometheus-exporter if { path /metrics }
   stats enable
   stats uri /stats
   stats refresh 10s"
fi

########## solr frontend
echo "
frontend ${HAPROXY_FRONTEND_SOLR:-solr}
  bind ${HAPROXY_FRONTEND_SOLR_BIND:-*:8880}
  default_backend ${HAPROXY_BACKEND_ALFRESCO:-alfresco}" 

########## alfresco backend
echo "
backend ${HAPROXY_BACKEND_ALFRESCO:-alfresco}
  option httpchk OPTIONS ${HAPROXY_BACKEND_ALFRESCO_CHECK:-/alfresco/s/api/server}
  server-template ${HAPROXY_BACKEND_ALFRESCO:-alfresco}- ${HAPROXY_BACKEND_ALFRESCO_COUNT:-1} ${HAPROXY_SERVICE_ALFRESCO:-alfresco}:8080 check resolvers docker init-addr libc,none"

########## resolvers
echo "
resolvers docker
  nameserver docker 127.0.0.11:53
  resolve_retries ${HAPROXY_RESOLVE_RETRIES:-3}
  timeout resolve ${HAPROXY_TIMEOUT_RESOLVE:-1s}
  timeout retry   ${HAPROXY_TIMEOUT_RETRY:-1s}
  hold other      ${HAPROXY_HOLD_OTHER:-10s}
  hold refused    ${HAPROXY_HOLD_REFUSED:-10s}
  hold nx         ${HAPROXY_HOLD_NX:-10s}
  hold timeout    ${HAPROXY_HOLD_TIMEOUT:-10s}
  hold valid      ${HAPROXY_HOLD_VALID:-10s}
  hold obsolete   ${HAPROXY_HOLD_OBSOLETE:-10s}" 


src/main/resources/templates/template-front-proxy-alfresco.sh
#!/bin/sh

set -e

HAPROXY_INCLUDE_WORKSPACE=${HAPROXY_INCLUDE_WORKSPACE:-false}
HAPROXY_INCLUDE_SHARE=${HAPROXY_INCLUDE_SHARE:-false}
HAPROXY_INCLUDE_FINDER=${HAPROXY_INCLUDE_FINDER:-false}
HAPROXY_INCLUDE_OOI=${HAPROXY_INCLUDE_OOI:-false}
HAPROXY_SEND_LOGS=${HAPROXY_SEND_LOGS:-false}
HAPROXY_INCLUDE_FRONTEND_STATS=${HAPROXY_INCLUDE_FRONTEND_STATS:-true}

########## global
echo "
global
  log-send-hostname ${HAPROXY_HOSTNAME:-localhost}
  log stdout format raw local0 debug
  maxconn 4096
  nbthread 1
  uid 99
  gid 99
  tune.ssl.default-dh-param 2048
  ssl-default-bind-ciphers ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES256-SHA384:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES256-SHA384:ECDHE-ECDSA-AES256-SHA:ECDHE-RSA-AES256-SHA:DHE-RSA-AES128-SHA256:DHE-RSA-AES128-SHA:DHE-RSA-AES256-SHA256:DHE-RSA-AES256-SHA:ECDHE-ECDSA-DES-CBC3-SHA:ECDHE-RSA-DES-CBC3-SHA:EDH-RSA-DES-CBC3-SHA:AES128-GCM-SHA256:AES256-GCM-SHA384:AES128-SHA256:AES256-SHA256:AES128-SHA:AES256-SHA:DES-CBC3-SHA:!DSS
  ssl-default-bind-options no-sslv3 no-tls-tickets
  stats socket /var/run/haproxy.sock mode 600 level admin
  stats timeout ${HAPROXY_STATS_TIMEOUT:-2m}"

if [[ $HAPROXY_SEND_LOGS = 'true' ]]
then
    echo "
  log ${HAPROXY_LOGS_SERVER:-10.88.10.50:514} len 2048 local0"
fi

########## defaults
echo "
defaults
  log global
  mode http
  cookie ${HAPROXY_COOKIE:-JSESSIONID} prefix nocache
  balance ${HAPROXY_BALANCE:-hdr(X-Forwarded-For)}
  retries 2
  timeout client ${HAPROXY_TIMEOUT_CLIENT:-30m}
  timeout connect ${HAPROXY_TIMEOUT_CONNECT:-4s}
  timeout server ${HAPROXY_TIMEOUT_SERVER:-30m}
  timeout check ${HAPROXY_TIMEOUT_CHECK:-5s}
  option redispatch
  option httplog"


########## stats frontend
if [[ $HAPROXY_INCLUDE_FRONTEND_STATS = 'true' ]]
then
echo "
frontend ${HAPROXY_FRONTEND_STATS:-stats}
   bind *:8404
   option http-use-htx
   http-request use-service prometheus-exporter if { path /metrics }
   stats enable
   stats uri /stats
   stats refresh 10s"
fi

########## internet frontend
echo "
frontend ${HAPROXY_FRONTEND_INTERNET:-internet}
  bind ${HAPROXY_FRONTEND_INTERNET_BIND:-*:80}
  default_backend ${HAPROXY_BACKEND_ALFRESCO:-alfresco}
  capture request header X-Forwarded-For len 15"
if [[ $HAPROXY_INCLUDE_SHARE = 'true' ]]
then
  echo "  use_backend ${HAPROXY_BACKEND_SHARE:-share} if { path_beg /share }" 
fi
if [[ $HAPROXY_INCLUDE_WORKSPACE = 'true' ]]
then
  echo "  use_backend ${HAPROXY_BACKEND_WORKSPACE:-workspace} if { path_beg /workspace }
  redirect code 301 prefix / drop-query append-slash if { path_reg ^/workspace$ }" 
fi
if [[ $HAPROXY_INCLUDE_FINDER = 'true' ]]
then
  echo "  use_backend ${HAPROXY_BACKEND_FINDER:-finder} if { path_beg /finder }"
fi
if [[ $HAPROXY_INCLUDE_OOI = 'true' ]]
then
  echo "  use_backend ${HAPROXY_BACKEND_OOI:-ooi-service} if { path_beg /ooi-service }"
fi

########## alfresco backend
echo "
backend ${HAPROXY_BACKEND_ALFRESCO:-alfresco}
  option httpchk OPTIONS ${HAPROXY_BACKEND_ALFRESCO_CHECK:-/alfresco/s/api/server}
  server-template ${HAPROXY_BACKEND_ALFRESCO:-alfresco}- ${HAPROXY_BACKEND_ALFRESCO_COUNT:-1} ${HAPROXY_SERVICE_ALFRESCO:-alfresco}:8080 check resolvers docker init-addr libc,none
  acl blacklist_alfresco path_sub /proxy/alfresco/api/solr/
  acl blacklist_alfresco path_sub /-default-/proxy/alfresco/api/
  acl blacklist_alfresco path_sub /service/api/solr/
  acl blacklist_alfresco path_sub /s/api/solr/
  acl blacklist_alfresco path_sub /wcservice/api/solr/
  acl blacklist_alfresco path_sub /wcs/api/solr/
  http-request deny if blacklist_alfresco"

########## share backend
if [[ $HAPROXY_INCLUDE_SHARE = 'true' ]]
then
echo "
backend ${HAPROXY_BACKEND_SHARE:-share}
  option httpchk OPTIONS ${HAPROXY_BACKEND_SHARE_CHECK:-/share}
  server-template ${HAPROXY_BACKEND_SHARE:-share}- ${HAPROXY_BACKEND_SHARE_COUNT:-1} ${HAPROXY_SERVICE_SHARE:-share}:8080 check resolvers docker init-addr libc,none"
fi

########## digital workspace backend
if [[ $HAPROXY_INCLUDE_WORKSPACE = 'true' ]]
then
echo "
backend ${HAPROXY_BACKEND_WORKSPACE:-workspace}
  option httpchk GET ${HAPROXY_BACKEND_WORKSPACE_CHECK:-/workspace}
  http-request replace-uri ([^\ :]*?)/workspace(.*) \1/\2
  server-template ${HAPROXY_BACKEND_WORKSPACE:-workspace}- ${HAPROXY_BACKEND_WORKSPACE_COUNT:-1} ${HAPROXY_SERVICE_WORKSPACE:-digital-workspace}:8080 check resolvers docker init-addr libc,none"
fi

########## finder backend
if [[ $HAPROXY_INCLUDE_FINDER = 'true' ]]
then
echo "
backend ${HAPROXY_BACKEND_FINDER:-finder}
  option httpchk GET ${HAPROXY_BACKEND_FINDER_CHECK:-/}
  http-request replace-uri ([^\ :]*?)/finder(.*) \1/\2
  server-template ${HAPROXY_BACKEND_FINDER:-finder}- ${HAPROXY_BACKEND_FINDER_COUNT:-1} ${HAPROXY_SERVICE_FINDER:-alfred-finder}:80 check resolvers docker init-addr libc,none"
fi

########## ooi-service backend
if [[ $HAPROXY_INCLUDE_OOI = 'true' ]]
then
echo "
backend ${HAPROXY_BACKEND_OOI:-ooi-service}
  server-template ${HAPROXY_BACKEND_OOI:-ooi-service}- ${HAPROXY_BACKEND_OOI_COUNT:-1} ${HAPROXY_SERVICE_OOI:-alfresco-ooi-service}:9095 check resolvers docker init-addr libc,none"
fi

########## resolvers
echo "
resolvers docker
  nameserver docker 127.0.0.11:53
  resolve_retries ${HAPROXY_RESOLVE_RETRIES:-3}
  timeout resolve ${HAPROXY_TIMEOUT_RESOLVE:-1s}
  timeout retry   ${HAPROXY_TIMEOUT_RETRY:-1s}
  hold other      ${HAPROXY_HOLD_OTHER:-10s}
  hold refused    ${HAPROXY_HOLD_REFUSED:-10s}
  hold nx         ${HAPROXY_HOLD_NX:-10s}
  hold timeout    ${HAPROXY_HOLD_TIMEOUT:-10s}
  hold valid      ${HAPROXY_HOLD_VALID:-10s}
  hold obsolete   ${HAPROXY_HOLD_OBSOLETE:-10s}"
